pyacq/devices/webcam_av.py

```diff
@@ -0,0 +1,83 @@
+"""
+av is python binding to libav or ffmpeg and this is so great (except the poor doc for the moment)
+http://mikeboers.github.io/PyAV/index.html
+"""
+
+
+import numpy as np
+
+from ..core import Node, register_node_type
+from pyqtgraph.Qt import QtCore, QtGui
+
+try:
+    import av
+    HAVE_AV = True
+except ImportError:
+    HAVE_AV = False
+
+import time
+
+
+class AVThread(QtCore.QThread):
+    def __init__(self, out_stream, container, parent=None):
+        QtCore.QThread.__init__(self)
+        self.out_stream = out_stream
+        self.container = container
+        
+    def run(self):
+        self.running = True
+        n = 0
+        stream = self.container.streams[0]
+        while (self.running):
+            for packet in self.container.demux(stream):
+                for frame in packet.decode():
+                    arr = frame.to_rgb().to_nd_array()
+                    n += 1
+                    self.out_stream.send(n, arr)
+
+
+class WebCamAV(Node):
+    """
+    Simple device that use the av python module.
+    See http://mikeboers.github.io/PyAV/index.html.
+    It is a wrapper on top ffmpeg or libav.
+    """
+    _output_specs = {'video' : dict(streamtype = 'video',dtype = 'uint8',
+                                                shape = (4800, 6400, 3), compression ='',
+                                                sampling_rate = 1.)
+                                }
+    def __init__(self, **kargs):
+        Node.__init__(self, **kargs)
+        assert HAVE_AV, "WebCamAV node depends on the `av` package, but it could not be imported."
+    
+
+    def configure(self, camera_num=0, **options):
+        self.camera_num = camera_num
+        self.options = options
+
+        container = av.open('/dev/video{}'.format(self.camera_num), 'r','video4linux2', self.options)
+        stream = next(s for s in container.streams if s.type == 'video')
+        self.output.spec['shape'] = (stream.format.height, stream.format.width, 3)
+        self.output.spec['sampling_rate'] = float(stream.average_rate)
+    
+    def _initialize(self):
+        pass
+    
+    def start(self):
+        self.container = av.open('/dev/video{}'.format(self.camera_num), 'r','video4linux2', self.options)
+
+        self._thread = AVThread(self.output, self.container)
+        self._thread.start()
+        self._running = True
+
+    def stop(self):
+        self._thread.running = False
```

Comment:
In general the GIL does not protect you from race conditions. Even though only one thread can run the python interpreter at a time, it is possible to get a crash if the interpreter locks in the middle of an operation in one thread, and then unlocks with a conflicting operation in another thread. 

In this case you _might_ be safe because the operation is so simple that it might be atomic from the standpoint of the GIL, but you are much better safe than sorry in cases like this, since the cost of being wrong is so high (in two years, when the code has become very complex, you might notice some random segmentation faults and wonder what cases them). I've been through this many times, and I have wasted weeks of time fixing these issues.

My personal rules to avoid threading-related crashes are: 
1. All threads are implemented as a class
2. Only the thread instance may access its own attributes; everyone else must go through a method
3. These methods must handle proper locking of attributes